hail/src/main/scala/is/hail/expr/ir/InferPType.scala

```diff
@@ -5,8 +5,44 @@ import is.hail.expr.types.virtual.{TNDArray, TTuple}
 import is.hail.utils._
 
 object InferPType {
+  private def getNestedElementPTypes(ptypes: Seq[PType]): PType = {
+    ptypes.head match {
+      case x: PStreamable => {
+        val elementType = getNestedElementPTypes(ptypes.map(_.asInstanceOf[PStreamable].elementType))
+        x.copyStreamable(elementType, ptypes.forall(_.required))
+      }
+      case _: PSet => {
+        val elementType = getNestedElementPTypes(ptypes.map(_.asInstanceOf[PSet].elementType))
+        PSet(elementType, ptypes.forall(_.required))
+      }
+      case x: PStruct => {
+        val allRequired = ptypes.forall(_.required)
+
+        PStruct(allRequired, x.fieldNames.map( fieldName =>
+          fieldName -> getNestedElementPTypes(ptypes.map(_.asInstanceOf[PStruct].field(fieldName).typ))
+        ):_*)
+      }
+      case x: PTuple => {
+        val allRequired = ptypes.forall(_.required)
+
+        PTuple(allRequired, x._types.map( pTupleField =>
+          getNestedElementPTypes(ptypes.map(_.asInstanceOf[PTuple]._types(pTupleField.index).typ))
+        ):_*)
+      }
+      case _: PDict => {
+        val allRequired = ptypes.forall(_.required)
+        val keyType = getNestedElementPTypes(ptypes.map(_.asInstanceOf[PDict].keyType))
+        val valueType = getNestedElementPTypes(ptypes.map(_.asInstanceOf[PDict].valueType))
+
+        PDict(keyType, valueType, allRequired)
+      }
+      case _ => ptypes.head.setRequired(ptypes.forall(_.required))
```

Comment:
need PInterval.

Reply:
How best to test requiredeness of PInterval values? I can set them to be all required, however, Literal( ) allows null values. In this case, my Literal case in InferPType could match on type, and for TInterval, examine the provided interval. Else, we can rely on the type passed to literal.